driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -0,0 +1,760 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import java.lang.reflect.ParameterizedType;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutionException;
+
+import com.google.common.base.Objects;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.TypeToken;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.datastax.driver.core.TypeCodec.*;
+import com.datastax.driver.core.exceptions.CodecNotFoundException;
+
+/**
+ * A registry for {@link TypeCodec}s. When the driver
+ * needs to serialize or deserialize an object,
+ * it will lookup in the {@link CodecRegistry} for
+ * a suitable codec.
+ *
+ * <h3>Usage</h3>
+ *
+ * <p>
+ * To build a {@link CodecRegistry} instance with all the default codecs used
+ * by the Java driver, simply use the following:
+ *
+ * <pre>
+ * CodecRegistry myCodecRegistry = new CodecRegistry().registerDefaults();
+ * </pre>
+ *
+ * Custom {@link TypeCodec}s can be added to a {@link CodecRegistry} via the {@code register} methods:
+ *
+ * <pre>
+ * CodecRegistry myCodecRegistry = new CodecRegistry()
+ *    .register(myCodec1, myCodec2, myCodec3)
```

Comment:
Would it be more advisable to recommend calling registerDefaults() first then register(myCodec...) so any custom registered codecs take precedence over the default ones?  Or will we not expect it to be normal to override default codecs?